### PR TITLE
add missing kube-api resources from managed agent manifest

### PR DESCRIPTION
**Pull request by oren-zohar**
## What does this PR do?
In order for cloudbeat `kube-api` fetcher to work properly, we need to update the agent-managed manifest with permissions related to kube-api resources.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Relates: https://github.com/elastic/elastic-agent/pull/179
- Relates: https://github.com/elastic/security-team/issues/3682

### CODE REVIEW

**Comment by ph** [comment]
@blakerouse or @ChrsMark Can you take a look?

**Comment by blakerouse** [comment]
I am going to have to defer to @ChrsMark, I do not know why it would need those permissions. Being that its related to that data-plane and not the control-plane (as Elastic Agent doesn't need that information for the the kubernetes dynamic input service).

File: deploy/kubernetes/elastic-agent-managed-kubernetes.yaml [hunk 1]
```diff
@@ -202,6 +203,22 @@ rules:
       - "/metrics"
     verbs:
       - get
+  - apiGroups: ["rbac.authorization.k8s.io"]
```
Comment (ChrsMark) [inline deploy/kubernetes/elastic-agent-managed-kubernetes.yaml:206]: These grant a lot of permissions to Agent's Pod by default. Is this something that we want for Agent? Wouldn't users that don't want to use the specific package/integration be worried seeing all these permissions?

I would think of adding them in an additional+seperate Role like what we do for the `leases` or `kubeadm-config` configmap. Also those could be dissabled by default and only grant them once the users actually want the specific package/integration. Let me know what you think about this.

Comment (oren-zohar) [reply deploy/kubernetes/elastic-agent-managed-kubernetes.yaml:206]: > These (and all changes in this PR in general) grant a lot of permissions to Agent's Pod by default. Is this something that we want for Agent? Wouldn't users that don't want to use the specific package/integration be worried seeing all these permissions?

We wanted the user to have cloudbeat working from the get-go, editing the deployment will add more friction to the process, but I see where you're coming from.

> I would think of adding them in an additional+seperate Role like what we do for the leases or kubeadm-config configmap. Also those could be dissabled by default and only grant them once the users actually want the specific package/integration. Let me know what you think about this.

I think splitting it into a separate role is a good idea. When you say "dissabled by default" you mean to commented out, right?

I think a good solution would be to keep our additional permission in a commented out role, and in case the user added our integration, we will promote the user with the same deployment manifest, but with our role enabled, @ChrsMark wdyt?

Comment (amitkanfer) [reply deploy/kubernetes/elastic-agent-managed-kubernetes.yaml:206]: @ChrsMark thank you!
I would like to provide my $0.02 on the additional permissions concern.
Today, whoever installs Elastic Agent, already allows read only access to a total of 11 resources: pods, namespaces, events, pods, services, configmaps, statefulsets, deplyments, replicasets, jobs and cronjobs.

The team is asking to extend this read-only access to 5 more resources - serviceaccounts, rolebindings, roles, clusterroles and clusterrolebindings. 

Important to emphasize that the ask is not to grant more permissive access to existing or new resources, but rather to extend the same level of access to more types of resources. The "extra permissions" in question are not intrusive in any way for the user. 

From a product point of view, if our users would have to redeploy the agent (with different permissions) for every new type of integration - this voids much of the value of the agent as a vehicle to easily deploy our integrations, as it's becoming not that easy anymore.

TL;DR - there's a tradeoff here, like in everything. I would like to make the case that to extend *read-only* access to more k8s resources compared to what we have today is a "risk" worth taking in order to reduce friction and to really make our deployment process as easy as possible for our users.

All of the above is not necessarily true to more intrusive types of permissions.

@ChrsMark wdyt?

/cc @norrietaylor @snehsach19 @tehilashn

Comment (norrietaylor) [reply deploy/kubernetes/elastic-agent-managed-kubernetes.yaml:206]: /cc @m-sample

Comment (ChrsMark) [reply deploy/kubernetes/elastic-agent-managed-kubernetes.yaml:206]: @amitkanfer I agree that these are read only permissions but still it is an additional exposure of cluster information and hence we should carefully consider it. Having it disabled by default and adding the Role on demand would work here since even if you have already deployed the Agent's Pod you can always assign a new role to it at any time without having to restart it, right? If we verify this for that specific case we can evaluate this option too.
However if we still want to invest more on the easier+faster onboarding of the user we can add those Roles in the basic manifest and we should be able to cover our user's questions in case they are cautious about this. These are the options we have here but I think this decision is to be answered by product better?

Comment (amitkanfer) [reply deploy/kubernetes/elastic-agent-managed-kubernetes.yaml:206]: @ChrsMark 
Thanks for the prompt response.
I think we all agree on the tradeoffs, and it's a good that this discussion happens and documented for our future selves. We should always carefully consider our product security practices. 

My opinion is that we should go above and beyond to reduce friction to the end user, especially in the onboarding / installation flows (like this one) even if it means we extend our read-only permissions to few more resources. My thoughts are different for more intrusive permissions.

I will let @hmnichols share his point of view from a product perspective.

Comment (ruflin) [reply deploy/kubernetes/elastic-agent-managed-kubernetes.yaml:206]: As these are all read only permissions (at least that is my understanding) I'm not too concerned around having these in by default. As mentioned before, it is a trade off.

What is important from my perspective is that a user that wants to remove these permissions has an easy way to do it. This does not only apply to the permissions added here but all permissions. Ideally docs exist that explain why which permissions are needed and what the consequences of removing these permissions are. If then someone still enables an integration that needs these removed permissions, we must have good error messages in place for the user to be able to debug why things don't work.

Comment (tonymeehan) [reply deploy/kubernetes/elastic-agent-managed-kubernetes.yaml:206]: @ruflin Thanks, and agreed.

@ChrsMark raises a good point, and I do think it's important for us to review and document the information contained in the resources we are providing read-only access to, both the 5 new resources in this PR (serviceaccounts, clusterrolebindings, clusterroles, rolebindings, roles), and perhaps the 11 we already grant by default as well. We should ensure we document what information is disclosed by providing read-only access to them in Agent.

Separately, we should come to an agreement on what the future of Agent looks like in Kubernetes, especially if we can take into account some of the points @ChrsMark raises here and employ components like operators to help us easily orchestrate the installation of integrations and guarantee least privilege for each integration in a way that's seamless/frictionless for the user.

**Recommendation** 

The combination of the following make me comfortable with merging this PR:

1. documenting the information contained in the resources
2. agreeing to a future where we can use things like operators to orchestrate the installation of integrations with least privilege

**Kube-API resources for K8s security posture management**

Regarding the information contained in these resources, @oren-zohar collected this information here:

These are the additional resources that are required in order to evaluate the cluster for compliance with the [CIS Kubernetes Benchmark v1.23](https://learn.cisecurity.org/l/799323/2022-02-09/qsd4y)

[Service Accounts](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#service-account)

- serviceAccountName
- automountServiceAccountToken

[Cluster Role bindings](https://kubernetes.io/docs/reference/kubernetes-api/authorization-resources/cluster-role-binding-v1/#ClusterRoleBinding)

- roleRef 
- roleRef.apiGroup
- roleRef.kind
- roleRef.name
- subjects
- subjects.kind
- subjects.name
- subjects.apiGroup
- subjects.namespace

[Role-bindings](https://kubernetes.io/docs/reference/kubernetes-api/authorization-resources/role-binding-v1/)
Relevant flags: authorization-mode

- aggregationRule
- aggregationRule.clusterRoleSelectors
- rules 
- rules.apiGroups
- rules.resources
- rules.verbs 
- rules.resourceNames
- rules.nonResourceURLs

[Roles](https://kubernetes.io/docs/reference/kubernetes-api/authorization-resources/role-v1/#Role)
Relevant fields: apiGroups, resources, verbs

- rules 
- rules.apiGroups
- rules.resources
- rules.verbs 
- rules.resourceNames
- rules.nonResourceURLs

[Pod Security policies](https://kubernetes.io/docs/concepts/security/pod-security-policy/)



Control Aspect | Field Names
-- | --
Running of privileged containers | privileged
Usage of host namespaces | hostPID, hostIPC
Usage of host networking and ports | hostNetwork, hostPorts
Usage of volume types | volumes
Usage of the host filesystem | allowedHostPaths
Allow specific FlexVolume drivers | allowedFlexVolumes
Allocating an FSGroup that owns the pod's volumes | fsGroup
Requiring the use of a read only root file system | readOnlyRootFilesystem
The user and group IDs of the container | runAsUser, runAsGroup, supplementalGroups
Restricting escalation to root privileges | allowPrivilegeEscalation, defaultAllowPrivilegeEscalation
Linux capabilities | defaultAddCapabilities, requiredDropCapabilities, allowedCapabilities
The SELinux context of the container | seLinux
The Allowed Proc Mount types for the container | allowedProcMountTypes
The AppArmor profile used by containers | annotations
The seccomp profile used by containers | annotations
The sysctl profile used by containers | forbiddenSysctls,allowedUnsafeSysctls

**Review by ChrsMark — APPROVED** [review]
We had a chat with @tonymeehan and it seems that we can proceed with this one for now and follow-up on documenting in detail why these permissions are needed and what happens if someone disable them. I think the most suitable place for this would be https://www.elastic.co/guide/en/fleet/master/running-on-kubernetes-managed-by-fleet.html.

Another thing that we need to think of is how we treat changes in those manifests and more specifically if a change in those manifests can be considered a breaking change for users. My  perspective is that so far we didn't treat those manifests the similar way we do with core codebase regarding breaking changes which means that we can remove/update specs/settings even if that breaks things.

**Comment by oren-zohar** [comment]
@ChrsMark I opened an issue to track the documentation part. Merging is blocked for me so if you can please grant me merging rights or merge it yourself that would be great.

**Review by ChrsMark — CHANGES_REQUESTED** [review]
This PR would require some extra changes so as to be merged:

1. I see that only the managed version is updated. Is this intentional? I would suggest ensuring same level of roles/permissions in both standalone and managed.
2. Files in https://github.com/elastic/elastic-agent/tree/main/deploy/kubernetes/elastic-agent-standalone and https://github.com/elastic/elastic-agent/tree/main/deploy/kubernetes/elastic-agent-managed should be updated too properly.  Actually changes should take place first in these files and then running [`make all`](https://github.com/elastic/elastic-agent/blob/main/deploy/kubernetes/Makefile) at top level should update the full files too. This should had been catched by our CI here, since it was like this in Beats repo. @narph do you know why this is not happening here?

**Review by ChrsMark — APPROVED** [review]
Overall looks good. However we need to make sure that process of updating these files is robust by using `make all` and that CI will ensure that no inconsistencies exist (cc: @narph )

File: version/docs/version.asciidoc [hunk 2]
```diff
@@ -1,4 +1,4 @@
-:stack-version: 8.0.0
+:stack-version: 8.2.0
```
[the remark this answers is not in the document]
Comment (narph) [reply version/docs/version.asciidoc]: this pr https://github.com/elastic/elastic-agent/pull/417 should update version and add check in CI

**Review by ChrsMark — APPROVED** [review]
